### DISCUSSION TITLE
A dev overlay

### DISCUSSION THREAD

**Discussion by Princesseuh** [withastro/roadmap#708] [Proposal]
### Body

# Summary

Tooling around Astro has certainly improved in the past two years, we have a working language server, some CLI tooling, ESLint and Prettier plugins etc. One thing we are missing however is help for the users during development. For this, I'd like to suggest a new feature: a dev overlay

# Background & Motivation

We've all had those moments where we run our website through a billion different online testers to make sure our website is performant, accessible and secure. But, those websites are very general, and can't necessarily give you instructions specific to your framework.

With a Astro specific dev overlay, we'd be able to give more precise Astro-specific infos, such as 

- Hey, maybe you should use the Image component here?
- This component is far below the fold, have you considered `client:visible`?
- etc

Additionally, it could provide tools that are also specific to Astro, such as the ability to visualize the islands on a page, or debug `getStaticPaths`.

# Goals

- Provide Astro users with helpful tools to debug and improve their website

## Non-goals

- Replace every tool out there. We'll never be as good as tools as Lighthouse, so we shouldn't be competing anyway.

**Comment by alexanderniebuhr** [comment]
First off, I love the idea of an Astro-specific dev overlay. It would be a game-changer in terms of developer experience. A few thoughts and questions:

It would be great to see support for Adapters. Perhaps the overlay could expose hooks or a robust API, allowing Adapters to implement specific features - which are only relevant to one Adapter - themselves?

Are you envisioning this overlay to be a part of the browser's dev tools, or injected, standalone (similar [Nuxt Devtools](https://devtools.nuxt.com/guide/features))?

One concern I have is about the overlay potentially affecting the site's code (CSS, JS, etc.). How do we ensure that the overlay, if injected into the site, doesn't interfere with or alter the original code?

**Comment by Princesseuh** [comment]
> First off, I love the idea of an Astro-specific dev overlay. It would be a game-changer in terms of developer experience. A few thoughts and questions:
> 
> It would be great to see support for Adapters. Perhaps the overlay could expose hooks or a robust API, allowing Adapters to implement specific features - which are only relevant to one Adapter - themselves?
> 
> Are you envisioning this overlay to be a part of the browser's dev tools, or injected, standalone (similar [Nuxt Devtools](https://devtools.nuxt.com/guide/features))?
> 
> One concern I have is about the overlay potentially affecting the site's code (CSS, JS, etc.). How do we ensure that the overlay, if injected into the site, doesn't interfere with or alter the original code?

Great questions!

Yes, in my mind Integrations and adapters would be able to add to it. I could imagine CloudFlare, Vercel etc wanting to add info very easily, for sure.

It would be injected, directly in the dev server. Developing extensions for every browser is fairly complicated at times and we'd love for this to work on all kinds of browsers including on mobile devices.

I'm thinking it would be implemented using a web component, similar to the error overlay, which allows CSS isolation. I've never heard of websites being bothered by the error overlay, so I assume this is a workable solution.

**Comment by Princesseuh** [comment]
This happened!
